Add FractionSum Calculation

brews opened the pull request:
This PR:
- Adds `FractionSum` calculation.
- Fixes a minor bug in `Production` exception message in `enable_deltamethod()`.

This PR includes docstrings and unit tests for the new feature.

Close #32.

brews commented:
Here is the simplest implementation I could think of. Feel free to rip and tear. It's ready for you to take a look at, @jrising .

One question, off the bat: I take in `frac` and `subcalcs` as initializing args. This is the first time I've had to split initializing subcalculations across two arguments. I think I have the `.describe()` right but my `.apply()` implementation feels hackish...?

jrising commented on openest/generate/functions.py:
```diff
@@ -443,6 +443,148 @@ def describe():
+    unshift : bool, optional
+    """
+
+    def __init__(self, frac, subcalcs, unshift=True):
```
The automatic argument matcher will get confused by splitting out the first of three subcalcs, and there's a second reason to keep them all in one list: you can generalize this to N subcalcs, by saying that they go in the following order:
`fraction1, subcalc1, [fraction2, subcalc2, ...], subcalcN`
where the fraction applied to `subcalcN` is just (1 - (fraction1 + fraction2 + ...)).

If you don't plan on setting this up in the general way, then I think you should just have three separate arguments:
`fraccalc`, `truecalc`, `falsecalc`
or similarly named.

brews replied:
Resolved with acc586a.

jrising commented on openest/generate/functions.py:
```diff
@@ -443,6 +443,148 @@ def describe():
+        fullunitses = subcalcs[0].unitses[:]
+        for ii in range(1, len(subcalcs)):
+            assert subcalcs[0].unitses[0] == subcalcs[ii].unitses[0], "%s <> %s" % (subcalcs[0].unitses[0], subcalcs[ii].unitses[0])
+            fullunitses.extend(subcalcs[ii].unitses)
```
Also check that frac is unitless.

brews replied:
@jrising What are accepted values for unitless units?

brews replied:
Resolved with ec0e39203139fd0632a354e398e6075aea9f8cfe

jrising commented on openest/generate/functions.py:
```diff
@@ -443,6 +443,148 @@ def describe():
+            alldeps.update(elements['main'].dependencies)
+
+        if lang in ['latex', 'julia']:
+            elements['main'] = FormatElement(' * '.join([main.repstr for main in mains]), list(alldeps))
```
Fix this.

brews replied:
We're actually raising a `NotImplementedError` before we get to this.

I'm not sure how to handle this for the Calculation.

brews replied:
Resolved with 26b0a1c.

jrising commented on openest/generate/functions.py:
```diff
@@ -443,6 +443,148 @@ def describe():
+        if lang in ['latex', 'julia']:
+            elements['main'] = FormatElement(' * '.join([main.repstr for main in mains]), list(alldeps))
+
+        formatting.add_label('product', elements)
```
Also here.

brews replied:
We're raising a NotImplementedError before we get to this.

brews replied:
Resolved with 26b0a1c.

jrising commented on openest/generate/functions.py:
```diff
@@ -443,6 +443,148 @@ def describe():
+        """
+        infoses = [subcalc.column_info() for subcalc in self.subcalcs]
+        title = 'FractionSum of previous results'
+        description = 'FractionSum of ' + ', '.join([infos[0]['title'] for infos in infoses])
```
This won't be very interpretable.

brews replied:
Does this need to be human readable or how is this information used/interpreted? Would you elaborate on how to handle this relatively sophisticated Calculation, in this case?

brews replied:
Resolved with 21cc3c0.

jrising commented on openest/generate/functions.py:
```diff
@@ -443,6 +443,148 @@ def describe():
+    def enable_deltamethod(self):
+        raise NotImplementedError
+
+    def partial_derivative(self, covariate, covarunit):
```
Any reason you're not implementing this?

brews replied:
`Product` isn't merged into `master` yet. I'll get on it as soon as it's in.

brews replied:
Resolved with 20ecad1, should be checked.

jrising commented on openest/test/generate/test_functions.py:
```diff
@@ -92,6 +92,85 @@ def test_spaninstabase():
+        """Tests that FractionSum correctly appends units from the original subcalcs
+        """
+        unit = ['fakeunit']
+        frac_mock = MockAppCalc(years=[0], values=[0.1], unitses=unit)
```
Shouldn't have `fakeunit`.

brews replied:
What do we use to indicate unitless-ness? Calculations require some form of input for `unitses`, so just a list of `None` (this is how it's now done in 
acc586a).

brews replied:
Resolved in ec0e39203139fd0632a354e398e6075aea9f8cfe. We're using "unitless".

jrising commented:
@brews The `describe` is fine, but I also felt like the `apply` was pretty cludgy. May make sense to define a new `Application` subclass that takes a *subcalcs and passes the *result[1]s to a waiting function. That's if you're going to stick to just supporting 3 subcalcs.

brews commented:
Note, I force pushed this after rebasing onto `master` to get the new `Production` class. We need this to implement the partial derivs method.

brews commented:
Ready for review, @jrising!

This PR has undergone a heavy rewrite following your feedback.

Please, especially check the new `.partial_derivative()` method - I'm not sure I handle the final weighted value correctly. You might be able to simplify the whole thing.

Thanks for your time. I know you're busy.

brews commented:
Thanks for the review, @jrising. Appreciate your patience on this.

- > Your 1 - {last_weight} math in format will fail if there are more than 2 terms.

Testing this I get:
```python
>>> subcalc_strs = ['a', 'b', 'c', 'd', 'e']  # b, d are unitless weight subcalcs
>>> last_weight = ' + '.join(subcalc_strs[1:-1:2])
>>> last_weight
'b + d'
>>> last_product = f"(1 - {last_weight}) * {subcalc_strs[-1]}"
>>> last_product
'(1 - b + d) * e'
```
Which looks good to me...? Let me know if I'm missing something obvious. I'll open a new issue and fix this if there is a bug.

- > I change the np.any(weights > 1) check in generate to ensure that sum(weights) < 1.

Changed in deb8e45.

- > Aren't you missing the un-multiplied self.subcalcs[-1] term in your partial derivative calculator?

Ah! Thank you. Fixed in e26227cb.

- > I think the complexity of setting up the expression in order to take its derivative defeats purpose of writing it that way.

I wouldn't fight you if you have any improvements. I mostly wanted to avoid dealing with the parts for the partials of all the products in the calculation.

jrising commented:
On 1, it should be '(1 - b - d) * e', unless I'm missing something obvious.

brews commented:
Man, that's sloppy of me. Thank you! Got the bug with PR #38
